include/shark/Data/BatchInterfaceAdaptStruct.h

```diff
@@ -292,114 +242,39 @@ public:\
-	SHARK_FUSION_DEFINE_STRUCT_INLINE(FusionShapeType, SHARK_TRANSFORM_BATCH_ATTRIBUTES(shape_type,ATTRIBUTES))\
-public:\
-	struct shape_type: public FusionShapeType{\
-		shape_type(){}\
-		template<typename... Args, typename = typename std::enable_if<!detail::is_variadic_constructible<shape_type, Args ...>::value>::type>\
-		shape_type(Args &&... args): FusionShapeType(std::forward<Args>(args)...){}\
-		template<class Archive>\
-		void serialize(Archive & archive,unsigned int version){\
-			boost::fusion::for_each(fusionize(), detail::ItemSerializer<Archive>(archive));\
-		}\
-		FusionShapeType& fusionize(){ return *this;}\
-		FusionShapeType const& fusionize() const{ return *this;}\
-	};\
-	struct type: public FusionType{\
-		typedef NAME value_type;\
-		template<typename... Args, typename = typename std::enable_if<!detail::is_variadic_constructible<type, Args ...>::value>::type >\
-		type(Args&&... args):FusionType(std::forward<Args>(args)...){}\
-		\
-		friend void swap(type& op1, type& op2){\
-			boost::fusion::swap(op1.fusionize(),op2.fusionize());\
-		}\
-		std::size_t size()const{\
-			return batchSize(boost::fusion::at_c<0>(fusionize()));\
-		}\
-		template<class Archive>\
-		void serialize(Archive & archive,unsigned int version){\
-			boost::fusion::for_each(fusionize(), detail::ItemSerializer<Archive>(archive));\
-		}\
-		FusionType& fusionize(){ return *this;}\
-		FusionType const& fusionize() const{ return *this;}\
-	};\
+#define SHARK_CREATE_BATCH_INTERFACE(NAME,ATTRIBUTES)\
+struct Batch< NAME >{\
 	typedef NAME value_type;\
-	typedef detail::TupleBatchReference<FusionRef, value_type> reference;\
-	typedef detail::TupleBatchReference<FusionConstRef, value_type> const_reference;\
-	\
-	static type createBatch(value_type const& input, std::size_t size = 1){\
-		type batch;\
-		boost::fusion::copy(boost::fusion::transform(input,detail::CreateBatch(size)),batch.fusionize());\
-		return batch;\
-	}\
-	template<class Iterator>\
-	static type createBatchFromRange(Iterator const& begin, Iterator const& end){\
-		std::size_t points = end - begin;\
-		type batch = createBatch(*begin,points);\
-		Iterator pos = begin;\
-		for(std::size_t i = 0; i != points; ++i,++pos){\
-			getBatchElement(batch,i) = *pos;\
-		}\
-		return batch;\
-	}\
-	static type createBatchFromShape(shape_type const& shape, std::size_t size = 1){\
-		type batch;\
-		detail::createFusionBatchFromShape<0>(batch.fusionize(),shape.fusionize(),size);\
-		return batch;\
-	}\
-	template<class T>\
-	static std::size_t size(T const& batch){return batch.size();}\
-	\
-	static reference get(type& batch, std::size_t i){\
-		return reference(batch, detail::MakeRef(i));\
-	}\
-	static const_reference get(type const& batch, std::size_t i){\
-		return const_reference(batch, detail::MakeConstRef(i));\
-	}
-
+	SHARK_CREATE_BATCH_INTERFACE_SEQ(SHARK_BATCH_MAKE_SEQUENCE(ATTRIBUTES))\
+};
+	
 #endif
```

Comment:
I don't only hate this obscure preprocessor magic, it is in a way far more intricate than templates, I also simply don't understand it. The conversions looks good though, as far as I can tell. However, this is an educated guess, at best.